check_upgrade/check_upgrade.go

```
@@ -0,0 +1,212 @@
+		        }
+	}`
+
+	resp, err := dg.NewTxn().Query(ctx, query)
+	if err != nil {
+		return []Node{}, errors.Wrapf(err, "while querying dgraph for duplicate nodes")
+	}
+
+	type Nodes struct {
+		Nodes []Node `json:"nodes"`
+	}
+	var result Nodes
+	err = json.Unmarshal([]byte(resp.Json), &result)
+	if err != nil {
+		return []Node{}, errors.Wrapf(err, "while unmarshalling response: %v", string(resp.Json))
+	}
+
+	duplicates := findDuplicateNodes(result.Nodes)
+
+	return duplicates, nil
+}
+
+func printDuplicates(entityType string, ns uint64, nodes []Node) {
+	if len(nodes) > 0 {
+		fmt.Printf("Found duplicate %ss in namespace: #%v\n", entityType, ns)
+		for _, node := range nodes {
+			fmt.Printf("dgraph.xid %v , Uid: %v\n", node.DgraphXID, node.UID)
+		}
+	}
+}
+
+func init() {
+	CheckUpgrade.Cmd = &cobra.Command{
+		Use:   "checkupgrade",
+		Short: "Run the checkupgrade tool",
+		Long:  "The checkupgrade tool is used to check for duplicate dgraph.xid's in the Dgraph database before upgrade.",
+		Run: func(cmd *cobra.Command, args []string) {
+			run()
+		},
+		Annotations: map[string]string{"group": "tool"},
+	}
+	CheckUpgrade.Cmd.SetHelpTemplate(x.NonRootTemplate)
+	flag := CheckUpgrade.Cmd.Flags()
+	flag.String(alphaGrpc, "127.0.0.1:9080",
+		"Dgraph Alpha gRPC server address")
+
+	flag.String(alphaHttp, "http://127.0.0.1:8080", "Draph Alpha HTTP(S) endpoint.")
+}
+
+func duplicateDeleteMutation(nodes []Node) {
+	if len(nodes) > 0 {
+		var deleteMutation strings.Builder
+		deleteMutation.WriteString("{\n")
+		deleteMutation.WriteString(" delete {\n")
+		for _, node := range nodes {
+			deleteMutation.WriteString(fmt.Sprintf("<%v> <%v> \"%v\" .\n", node.UID, "dgraph.xid", node.DgraphXID))
+		}
+		deleteMutation.WriteString(" }\n")
+		deleteMutation.WriteString(" }\n")
+
+		fmt.Printf("\ndelete duplicate nodes using following mutation : \n%v", deleteMutation.String())
+	}
+}
+
+func run() {
+	fmt.Println("Running check-upgrade tool")
```

This function should be called something else and should return an error so that we do not need to do `fmt.Fprintln(os.Stderr, err)` everywhere. And then, `run` function should call this function

---

We should write a test for this function too